torch/csrc/distributed/c10d/TCPStore.cpp

```diff
@@ -30,6 +30,51 @@
 namespace c10d {
 namespace detail {
 
+class timing_guard {
+  Counter& counter_;
+  typedef std::chrono::time_point<std::chrono::high_resolution_clock>
+      time_point;
+  time_point start_;
+
+ public:
+  timing_guard(Counter& counter)
+      : counter_(counter), start_(std::chrono::high_resolution_clock::now()) {}
+
+  ~timing_guard() {
+    stop();
+  }
+
+  void stop() {
+    if (start_ != time_point()) {
+      auto diff = std::chrono::duration_cast<std::chrono::milliseconds>(
+                      std::chrono::high_resolution_clock::now() - start_)
+                      .count();
+      counter_.update(diff);
+      start_ = time_point();
+    }
+  }
+};
+
+void Counter::update(double val) {
+  count_ += 1;
+
+  auto delta = val - mean_;
+  mean_ += delta / count_;
+
+  auto delta2 = val - mean_;
+  m2_ += delta2 * delta2;
```

Comment:
is `delta2^2` the right formula here?
https://www.johndcook.com/blog/standard_deviation/

would it be more like
```
      old_mean = mean_;
      mean_ += delta / count_;
      m2_  += (val - old_mean)*(val - mean_);
```

Reply:
The wikipedia article says that the original method from Knuth is numerically unstable and this version addresses some of those issues. https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm

It doesn't help that this is a [[citation needed](https://en.wikipedia.org/wiki/Wikipedia:Citation_needed)] remark.

I've used the second method successfully before.